core/domain/story_services.py

```diff
@@ -575,6 +575,33 @@ def get_completed_node_ids(user_id, story_id):
     return progress_model.completed_node_ids if progress_model else []
 
 
+def get_latest_completed_node_ids(user_id, story_id):
+    """Returns the ids of the completed nodes that come latest in the story.
+
+    Args:
+        user_id: str. ID of the given user.
+        story_id: str. ID of the story.
+
+    Returns:
+        list(str). List of the completed node ids that come latest in the story.
+            If length is larger than 3, return the last three of them.
+            If length is smaller or equal to 3, return all of them.
+    """
+    progress_model = user_models.StoryProgressModel.get(
+        user_id, story_id, strict=False)
+
+    if not progress_model:
+        return []
+
+    num_of_nodes = min(len(progress_model.completed_node_ids), 3)
+    completed_node_ids = [
+        node_id for node_id in progress_model.completed_node_ids]
+    completed_node_ids.sort(
+        key=story_domain.StoryNode.get_number_from_node_id)
+
+    return completed_node_ids[-num_of_nodes:]
```

Comment:
Hi @aks681 Could you explain a little more about how story nodes work? What would happen if the creator set the third node as the initial node? Why does that make the last 3 nodes node_1, node_4, node_5? I thought the node ids must be incremented.

Reply:
In the story editor, when the creator creates the nodes, their IDs are indeed created sequentially. So, 3 nodes created successively would be node_1, node_2 and node_3.
But after that, there is no restriction on the user to make, let's say node_3, as the initial node.
So, then the user can make the graph such that it would be node_3 -> node_2 -> node_1.

I'm just saying that this is a possibility as we don't have any checks that prevents the user from doing this.

Reply:
@aks681 To answer your question, it should be based strictly on the graph traversal, not the order of completion by the learner. (Sorry, I had actually missed when reviewing that the node ID numbers may not correspond to the former.)

@aks681 in your PR I saw story.story_contents.get_ordered_nodes, would that help here?

Reply:
Yeah, that function could be reused here.
Also, do you mean that it shouldn't be the order of creation by the creator? As order of completion by learner is the way in which the graph would be traversed, right?

Reply:
Neither of the above -- it should be the order specified by the creator based on the collection path.

(Imagine a case where the collection author arranges the graph one way, a learner completes it partially, then the order of the nodes is changed. In that case the two orders in your last question are not the same.)

Reply:
Done. I made use of get_ordered_nodes function in the PR, but I didn't include the backend test because I assume the PR will come in soon as well.